### ISSUE TITLE
JSON Parameter in API always empty

### ISSUE THREAD

**Issue by zeymermontes** [FlutterFlow/flutterflow-issues#2252]
### Has your issue been reported?

- [X] I have searched the existing issues and confirm it has not been reported.
- [X] I give permission for members of the FlutterFlow team to access and test my project for the sole purpose of investigating this issue.

### Current Behavior

I have an API with a variable of type json. The value is set with a Datatype as JSON. The request is sending this variable as empty json.
I printed the DataType as JSON and it is set correctly

![image](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/0b28067b-77a8-4f93-b30d-44a78da93468/b8c9e98f-b2cb-44a3-a3a9-78db5d8dacbb?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS8wYjI4MDY3Yi03N2E4LTRmOTMtYjMwZC00NGE3OGRhOTM0NjgvYjhjOWU5OGYtYjJjYi00NGEzLWEzYTktNzhkYjVkOGRhY2JiIiwiaWF0IjoxNzA3MTU2Mzg0LCJleHAiOjMzMjc3NzE2Mzg0fQ.\_ZwnfRe2d93hmNO_WmyXUU5wKVnLMyVzqOAydji7vXA)

![image](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/53aa1220-eaee-4374-89e7-37249c09139e/4d3434cb-bec5-4b24-8905-08bb14861bb9?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS81M2FhMTIyMC1lYWVlLTQzNzQtODllNy0zNzI0OWMwOTEzOWUvNGQzNDM0Y2ItYmVjNS00YjI0LTg5MDUtMDhiYjE0ODYxYmI5IiwiaWF0IjoxNzA3MTU2Mzg0LCJleHAiOjMzMjc3NzE2Mzg0fQ.OU88Db8yH3aWrDCnAUsQLOt6ZS7qgIcg_vKZortN3z0)

![image](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/2a3824af-e6a3-44fb-b8db-8fc620b57ec8/fd913921-849c-4af6-b265-73cee2eddc64?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS8yYTM4MjRhZi1lNmEzLTQ0ZmItYjhkYi04ZmM2MjBiNTdlYzgvZmQ5MTM5MjEtODQ5Yy00YWY2LWIyNjUtNzNjZWUyZWRkYzY0IiwiaWF0IjoxNzA3MTU2Mzg0LCJleHAiOjMzMjc3NzE2Mzg0fQ.8imcL3biIjqM6t8Yh9wiZVcQzE06rTBpqRX1P05cwXQ)

![image](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/c026f61b-76ac-4c25-a24c-69bf264ea1fa/2a0bfb85-100e-4899-b0ae-4721c2e73e79?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS9jMDI2ZjYxYi03NmFjLTRjMjUtYTI0Yy02OWJmMjY0ZWExZmEvMmEwYmZiODUtMTAwZS00ODk5LWIwYWUtNDcyMWMyZTczZTc5IiwiaWF0IjoxNzA3MTU2Mzg0LCJleHAiOjMzMjc3NzE2Mzg0fQ.uoSrRgO5KyHRMmM4pNank9qGXYq9hzNYmPbaHoG0TZ4)

![image](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/00bd4739-9fd8-4c0a-b721-2a9f0e25e837/66a89801-e86a-4c94-b9fd-6e105105973a?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS8wMGJkNDczOS05ZmQ4LTRjMGEtYjcyMS0yYTlmMGUyNWU4MzcvNjZhODk4MDEtZTg2YS00Yzk0LWI5ZmQtNmUxMDUxMDU5NzNhIiwiaWF0IjoxNzA3MTU2Mzg0LCJleHAiOjMzMjc3NzE2Mzg0fQ.ocb8xznfQwyq738swFOv-shJSgW0WQt26itMXVnll1w)

### Expected Behavior

I would expect the API call to have the variable with the correct JSON Value.

### Steps to Reproduce

1. Create an api call with a json variable
2. Set that variable from a DataType as JSON

### Reproducible from Blank

- [X] The steps to reproduce above start from a blank project.

### Bug Report Code (Required)

IT4glfLq25BOxPVA7c/+be9amiQXG1g7Roc0st5Hd0gdCOPrBuwuY+PsblBBZ9+maVNcJlL9jmYFpODOkd/DMMAnOk6aQJR5wpZiSwL0e2inaYTXBs+OekFPN8RjCBWuy5ivqhV7HMV2WiUa6173DOvIbDfDFOu/Zwh9f6PHaOI=

### Context

I need to send metadata to Stripe. Thats the only way I can relate each transaction with the order document

### Visual documentation

![image](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/a5f0b4fa-b56b-431b-a2ab-7c41230e97bd/de4efb62-c934-4e0c-92f1-9c1df9be4e06?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS9hNWYwYjRmYS1iNTZiLTQzMWItYTJhYi03YzQxMjMwZTk3YmQvZGU0ZWZiNjItYzkzNC00ZTBjLTkyZjEtOWMxZGY5YmU0ZTA2IiwiaWF0IjoxNzA3MTU2Mzg0LCJleHAiOjMzMjc3NzE2Mzg0fQ.D5B-p0WpHpVrW2YncxfNEmgXt0rvYIoQM6ZZQPuGeKk)

![image](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/cb52b39b-4b1f-4923-b25e-ede5b1982477/9959fb09-6c1c-4224-a0c7-6f06dfc89a98?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS9jYjUyYjM5Yi00YjFmLTQ5MjMtYjI1ZS1lZGU1YjE5ODI0NzcvOTk1OWZiMDktNmMxYy00MjI0LWEwYzctNmYwNmRmYzg5YTk4IiwiaWF0IjoxNzA3MTU2Mzg0LCJleHAiOjMzMjc3NzE2Mzg0fQ.Y5lbExmboRX-nCU8jEgbAcin1WpaccDEjb188jFxufw)

![image](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/62cf83d9-dd1b-4c4f-9f59-d9a87b0c1ad4/24b90337-75c0-4511-a827-81cd9d910261?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS82MmNmODNkOS1kZDFiLTRjNGYtOWY1OS1kOWE4N2IwYzFhZDQvMjRiOTAzMzctNzVjMC00NTExLWE4MjctODFjZDlkOTEwMjYxIiwiaWF0IjoxNzA3MTU2Mzg0LCJleHAiOjMzMjc3NzE2Mzg0fQ.FwseHkLiEvgOepuAtFvFvNSkWLAmOSDCEpezsmfopHA)

![image](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/46476898-3942-454c-904d-817236e41932/e236035a-4d9b-43ee-b366-4b4c42e012b4?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS80NjQ3Njg5OC0zOTQyLTQ1NGMtOTA0ZC04MTcyMzZlNDE5MzIvZTIzNjAzNWEtNGQ5Yi00M2VlLWIzNjYtNGI0YzQyZTAxMmI0IiwiaWF0IjoxNzA3MTU2Mzg0LCJleHAiOjMzMjc3NzE2Mzg0fQ.6ZGPieIyi02fCVcA-x10obKWf2LV3oKwZPAffmJeCEo)

![image](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/46e290b6-6e8d-4266-9b4f-ecfdb0385af3/f21638e1-d488-4575-9cdf-d27d3e610cab?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS80NmUyOTBiNi02ZThkLTQyNjYtOWI0Zi1lY2ZkYjAzODVhZjMvZjIxNjM4ZTEtZDQ4OC00NTc1LTljZGYtZDI3ZDNlNjEwY2FiIiwiaWF0IjoxNzA3MTU2Mzg0LCJleHAiOjMzMjc3NzE2Mzg0fQ.fRV61NBvgVNjqPrXOJn2jBefhgk0S7fN_s00vmzArYw)

### Additional Info

*No response*

### Environment

```
- FlutterFlow version: 4.1
- Platform: MacOS
- Browser name and version: Chrome Version 120.0.6099.129
- Operating system and version affected: Web
```

**Comment by nickydlt** [comment]
Is this issue going to get any attention? @leighajarett @hariprasadms @sbis04 ?

**Comment by leighajarett** [comment]
Looking into it! Sorry for the delay, thanks for understanding

**Comment by nickydlt** [comment]
Thank you, @leighajarett !

**Comment by nickydlt** [comment]
How's it going, @hariprasadms ?

**Comment by nickydlt** [comment]
@leighajarett & @hariprasadms ... 😳 ... getting this issue resolved is very important. Please provide a status update or some sign of life. Thank you!

**Comment by hariprasadms** [comment]
Hi @nickydlt -Thanks for reporting the issue. I've checked your project and the endpoint mentioned, but couldn't locate the configuration you referred to. It seems you may have made changes recently, which is fine.

To help investigate further, could you please provide the following details:

1. The actual endpoint: Does it have a meta data property in the post request?
2. What is the value of the meta data you're sending in the post request?
3. How does it work in Postman or any equivalent API client? Please share a sample post request and response.

This additional information will assist us in diagnosing the issue more accurately. Thanks!

**Comment by nickydlt** [comment]
@hariprasadms I needed to temporarily remove metadata parameters from the Body of "OneTimePublisherPayment" , "apostlePaymentOrSubscription" , and "publisherSusbcriptions". I've added them back in so you can look into this. Please check into this as quickly as possible because until it is resolved, the transactions on our app will not work properly. Thank you!

**Comment by hariprasadms** [comment]
Could you add answers above questions please.  As per my understanding the stripe checout end-point does not take meta data and it is optional. Did you test this in postman ?

**Comment by nickydlt** [comment]
I'll let @zeymermontes answer - I'm not smart enough lol.

**Comment by hariprasadms** [comment]
@nickydlt - Meanwhile i would suggest reach out FlutteFlow support team to get help around this kind of issues. However I would like see to what is that meta data response in actual api repose. Thanks

**Comment by hariprasadms** [comment]
Hi , What's the latest on the approach I suggested earlier? If there are no further updates, I assume the issue has been resolved on your end. If you still believe there's an issue, I recommend reaching out to the FF support team first to check the API configurations. If they confirm it's an issue, then feel free to reopen it here. Thanks

**Comment by zeymermontes** [comment]
The issue is well explained on the images.

Yes the endpoint has a matadata property: Its not one its multiple endpoints from stripe.

<img width="619" alt="image" src="https://github.com/FlutterFlow/flutterflow-issues/assets/66459565/0062d680-e7cd-41d8-9575-aeb029eac66f">

The value of metadata is this one(the second printed value):

![image](https://github.com/FlutterFlow/flutterflow-issues/assets/66459565/3cf92e7a-ab63-418c-82f5-f15cddfe0a1e)

The issue can't be replicated on Postman, because FF is sending the field metadata empty. The only way to reproduce is by manually sending meta data empty.

this is the payload as is being sent from FF:

![image](https://github.com/FlutterFlow/flutterflow-issues/assets/66459565/a1e1d62b-5f4d-4acf-9837-1ff5836c71fb)

**Comment by hariprasadms** [comment]
Please answer for my questions sent earlier. Those details are very much required to investigate the issue. What is the actual output you get in postman. Please add these details.

**Comment by zeymermontes** [comment]
Give me a minute

**Comment by zeymermontes** [comment]
<img width="902" alt="image" src="https://github.com/FlutterFlow/flutterflow-issues/assets/66459565/42f65214-13a5-4024-93d2-f347bcb24e24">

**Comment by takashiff** [comment]
Hi @zeymermontes, thank. you for the bug report. It looks like you are using a custom function to convert the DataType into a JSON, which is causing issues. We natively support DataType->JSON conversion, so I would recommend using that instead. Thanks

**Comment by zeymermontes** [comment]
I was using the native feature, thats the one I used to print the value on the image

![image](https://github.com/FlutterFlow/flutterflow-issues/assets/66459565/7e8c985e-0d42-4021-a5d3-0a2b55da7fd2)

**Comment by zeymermontes** [comment]
Then I created a custom function and the result was the same.

**Comment by takashiff** [comment]
@zeymermontes Got it, I'll continue looking into the issue

**Comment by apppbuilder** [comment]
Any point on this? @hariprasadms

**Comment by apppbuilder** [comment]
It's been days... Anyone?  @leighajarett @hariprasadms @sbis04

**Comment by leighajarett** [comment]
So sorry for the delay here, our engineering team is working on it!

**Comment by zeymermontes** [comment]
@leighajarett  thanks for the update.

**Comment by apppbuilder** [comment]
@leighajarett awesome!

**Comment by rzambroni** [comment]
Hey @apppbuilder our team is having a bit of a hard time replicating this. We suspect there might be an issue with your function dataTypeToJson(). Could you try the following: 

1) Use the FlutterFlow built-in json conversion or
2) Confirm their custom function is outputting valid json?

![image](https://github.com/FlutterFlow/flutterflow-issues/assets/54465214/6af698ab-d9fa-4751-a773-8d8eb09d9a33)

**Comment by zeymermontes** [comment]
@rzambroni  I used both  the buit in feature and my custom one, it was the same error

**Comment by glgh** [comment]
I'm looking into the issue but had trouble reproducing the error. Without further information, the best guess I have is there are some issues with the way we serialize DocumentReference, and as a result the json object wasn't returned properly. 

@zeymermontes can you please try the following and let me know if either changes the outcome?
- Temporarily remove productReference and userReference from stripeMetadata, and see if the API metadata is populated. This will help us narrow down where the problem is occurring.
- In your custom function dataTypeToJson(), wrap the return value 'json' with jsonEncode()

**Comment by apppbuilder** [comment]
@glgh @rzambroni Sorry for being late, I'm currently finishing another step inside my project, but I'll finish it today, then I can do the tests and reply everything.

Also, I believe that the problem is, as @glgh said, in the way you all serialize DocumentReference because I'm having the same problem in serializing the JSON output in DataType document that contains a DocumentReference, if I remove the docRef everything works (while receiving and parsing the data from the Cloud Function API instead of parsing and sending it)

![error](https://github.com/FlutterFlow/flutterflow-issues/assets/160440373/fa6b380a-8173-4599-a6ee-db292bdf0829)

I don't know if this infers in the enconding problem and because of that I'll open a new ticket, but until the end of the day I'll check the "To JSON" and give you all a feedback.

**Comment by apppbuilder** [comment]
Ok, here we go.

I made a controled test to check everything now. The good thing is: **It's working, but not fully working.**

I'll show you all below:
![bodyAPI](https://github.com/FlutterFlow/flutterflow-issues/assets/160440373/f0d37726-4d96-4863-a016-f3aff36c9357)
![bodyAPIVariables](https://github.com/FlutterFlow/flutterflow-issues/assets/160440373/6b515608-23a6-4d72-af26-4e742585ea44)
![words_learnedStruct](https://github.com/FlutterFlow/flutterflow-issues/assets/160440373/7f0cd237-08ec-4e6d-bc0d-c8f55fad5fd3)
![historyToReviewStruct](https://github.com/FlutterFlow/flutterflow-issues/assets/160440373/567a2461-7f13-473a-be35-e38d52ab035f)
![DocFromAuthUserCollecWithoutDocRef](https://github.com/FlutterFlow/flutterflow-issues/assets/160440373/b0371394-237b-44d9-bfbb-9372d29ec173) (from App State)
![DocFromAppStateWithDocRef](https://github.com/FlutterFlow/flutterflow-issues/assets/160440373/4cf8283a-7fd6-4174-87c0-39b571294e5e) (from Firebase User collection)

![LogsToCheckConsole1](https://github.com/FlutterFlow/flutterflow-issues/assets/160440373/707a11ef-f282-4513-8299-2984fbaaa7f0)

### And the results:
![shownDataInCloudResultsLog](https://github.com/FlutterFlow/flutterflow-issues/assets/160440373/825564c2-9977-486f-8d34-e325365a7a6f)

So yeah, there's a high chance it's in the way document Reference is being serialized...

**Comment by apppbuilder** [comment]
Any point on this? @rzambroni @glgh

**Comment by leighajarett** [comment]
This is in progress, we're currently working on it

**Comment by glgh** [comment]
We have identified the bug - working on a fix now.

**Comment by apppbuilder** [comment]
Hey everyone! @glgh is there a date to deploy the fix?

I just made the same call I've sent in the test and the documentReference is still brokening and not sending the whole data.

**Comment by glgh** [comment]
@apppbuilder the fix is in the pipeline, hopefully we will release it soon once we've done testing.

As a temporary workaround, I believe you can define a custom function **that only serializes the "path" part of any DocumentReference** (basically do exactly as you posted here https://github.com/FlutterFlow/flutterflow-issues/assets/16108776/0ebf0944-2f74-4cf4-a888-30db035f5678)

**Comment by apppbuilder** [comment]
@glgh Thank you for your reply.

Also, just to check:

Does this bug fix will impact this opened ticket too? #2842 **JSON data with DocumentReference from Firebase not being decoded.**

> Here we talking about enconding the DocumentReference, the ticket mentioned above is about DEcoding the data that contains DocumentReference.
